### ISSUE TITLE
Out of memory on ec2 instance

### ISSUE THREAD

**Issue by yzhang-myob** [loadimpact/k6#328]
Hi guys,

I have a test script which runs fine on local MAC OS, but gives out of memory error when executing on Amazon Linux of a r4.large instance. Can someone point out where I should look at?

Thank you

```
k6 run -s 5m:8000 -s 5m -s 5m:1000 src/client-credentials.js

          /\      |‾‾|  /‾‾/  /‾/
     /\  /  \     |  |_/  /  / /
    /  \/    \    |      |  /  ‾‾\
   /          \   |  |‾\  \ | (_) |
  / __________ \  |__|  \__\ \___/  Welcome to k6 v0.17.1!

  execution: local
     output: -
     script: /home/ec2-user/workspace/ids_load_tests_sit/src/client-credentials.js (js)

   duration: 0s, iterations: 0
        vus: 1, max: 8000

    web ui: http://127.0.0.1:6565/

fatal error: runtime: out of memory

runtime stack:
runtime.throw(0xb4750e, 0x16)
	/usr/local/Cellar/go/1.8.3/libexec/src/runtime/panic.go:596 +0x95
runtime.sysMap(0xc4f5de0000, 0x100000, 0x44f300, 0x100a838)
	/usr/local/Cellar/go/1.8.3/libexec/src/runtime/mem_linux.go:216 +0x1d0
runtime.(*mheap).sysAlloc(0xff13a0, 0x100000, 0x7fd25b484270)
	/usr/local/Cellar/go/1.8.3/libexec/src/runtime/malloc.go:440 +0x374
runtime.(*mheap).grow(0xff13a0, 0x1, 0x0)
	/usr/local/Cellar/go/1.8.3/libexec/src/runtime/mheap.go:774 +0x62
runtime.(*mheap).allocSpanLocked(0xff13a0, 0x1, 0x7fd25b484270)
	/usr/local/Cellar/go/1.8.3/libexec/src/runtime/mheap.go:678 +0x44f
runtime.(*mheap).alloc_m(0xff13a0, 0x1, 0xd, 0x7fd25b45d000)
	/usr/local/Cellar/go/1.8.3/libexec/src/runtime/mheap.go:562 +0xe2
runtime.(*mheap).alloc.func1()
	/usr/local/Cellar/go/1.8.3/libexec/src/runtime/mheap.go:627 +0x4b
runtime.systemstack(0x7fff73681480)
	/usr/local/Cellar/go/1.8.3/libexec/src/runtime/asm_amd64.s:343 +0xab
runtime.(*mheap).alloc(0xff13a0, 0x1, 0x1000000000d, 0x7fd25b484270)
	/usr/local/Cellar/go/1.8.3/libexec/src/runtime/mheap.go:628 +0xa0
runtime.(*mcentral).grow(0xff2c80, 0x0)
	/usr/local/Cellar/go/1.8.3/libexec/src/runtime/mcentral.go:212 +0x8e
runtime.(*mcentral).cacheSpan(0xff2c80, 0x7fd25b484270)
	/usr/local/Cellar/go/1.8.3/libexec/src/runtime/mcentral.go:93 +0x12a
runtime.(*mcache).refill(0x7fd2613a1000, 0xc40000000d, 0x7fd25b484270)
	/usr/local/Cellar/go/1.8.3/libexec/src/runtime/mcache.go:122 +0xa0
runtime.(*mcache).nextFree.func1()
	/usr/local/Cellar/go/1.8.3/libexec/src/runtime/malloc.go:538 +0x32
runtime.systemstack(0xfeb600)
	/usr/local/Cellar/go/1.8.3/libexec/src/runtime/asm_amd64.s:327 +0x79
runtime.mstart()
	/usr/local/Cellar/go/1.8.3/libexec/src/runtime/proc.go:1132

goroutine 1 [running]:
runtime.systemstack_switch()
	/usr/local/Cellar/go/1.8.3/libexec/src/runtime/asm_amd64.s:281 fp=0xc4237c4f20 sp=0xc4237c4f18
runtime.(*mcache).nextFree(0x7fd2613a1000, 0xd, 0x18, 0xaa5ac0, 0x10)
	/usr/local/Cellar/go/1.8.3/libexec/src/runtime/malloc.go:539 +0xb9 fp=0xc4237c4f78 sp=0xc4237c4f20
runtime.mallocgc(0xc0, 0xad46a0, 0xc4f5da6a01, 0xc4f5dccd80)
	/usr/local/Cellar/go/1.8.3/libexec/src/runtime/malloc.go:691 +0x827 fp=0xc4237c5018 sp=0xc4237c4f78
runtime.newobject(0xad46a0, 0xc4f5dccd80)
	/usr/local/Cellar/go/1.8.3/libexec/src/runtime/malloc.go:820 +0x38 fp=0xc4237c5048 sp=0xc4237c5018
github.com/dop251/goja.(*Runtime).newFunc(0xc4f5bc0180, 0xc4240972ec, 0x3, 0x1, 0xe65001, 0xc4f5dccd40)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/runtime.go:317 +0x8b fp=0xc4237c50f8 sp=0xc4237c5048
github.com/dop251/goja.(*newFunc).exec(0xc422134d50, 0xc4f4ced040)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/vm.go:1952 +0x60 fp=0xc4237c5178 sp=0xc4237c50f8
github.com/dop251/goja.(*vm).run(0xc4f4ced040)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/vm.go:288 +0x7f fp=0xc4237c51b0 sp=0xc4237c5178
github.com/dop251/goja.(*funcObject).Call(0xc4f5c02a80, 0xe65040, 0xc4f5dcc240, 0xc4f5c062a0, 0x3, 0x1d4, 0xfe1750, 0xc4237c52c8)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/func.go:130 +0x324 fp=0xc4237c5258 sp=0xc4237c51b0
github.com/dop251/goja.(*funcObject).Call-fm(0xe65040, 0xc4f5dcc240, 0xc4f5c062a0, 0x3, 0x1d4, 0x4085ce, 0xa45f40)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/builtin_function.go:120 +0x51 fp=0xc4237c52a8 sp=0xc4237c5258
github.com/dop251/goja.(*Runtime).functionproto_call(0xc4f5bc0180, 0xe65040, 0xc4f5bf8f00, 0xc4f5c06290, 0x4, 0x1d5, 0xb37edd, 0xc4237c53c0)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/builtin_function.go:74 +0x157 fp=0xc4237c5340 sp=0xc4237c52a8
github.com/dop251/goja.(*Runtime).(github.com/dop251/goja.functionproto_call)-fm(0xe65040, 0xc4f5bf8f00, 0xc4f5c06290, 0x4, 0x1d5, 0xc4f5bbd620, 0x40b222)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/builtin_function.go:160 +0x51 fp=0xc4237c5390 sp=0xc4237c5340
github.com/dop251/goja.(*vm)._nativeCall(0xc4f4ced040, 0xc4f5bc4000, 0x4)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/vm.go:1827 +0x2e9 fp=0xc4237c5438 sp=0xc4237c5390
github.com/dop251/goja.call.exec(0xc400000004, 0xc4f4ced040)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/vm.go:1808 +0x6ff fp=0xc4237c5500 sp=0xc4237c5438
github.com/dop251/goja.(*call).exec(0xc42110f840, 0xc4f4ced040)
	<autogenerated>:816 +0x55 fp=0xc4237c5540 sp=0xc4237c5500
github.com/dop251/goja.(*vm).run(0xc4f4ced040)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/vm.go:288 +0x7f fp=0xc4237c5578 sp=0xc4237c5540
github.com/dop251/goja.(*funcObject).Call(0xc4f5c029c0, 0xe65040, 0xc4f5dcc1e0, 0xc4f5c061a0, 0x3, 0x1e4, 0xfe1750, 0xc4237c5690)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/func.go:130 +0x324 fp=0xc4237c5620 sp=0xc4237c5578
github.com/dop251/goja.(*funcObject).Call-fm(0xe65040, 0xc4f5dcc1e0, 0xc4f5c061a0, 0x3, 0x1e4, 0x4085ce, 0xa45f40)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/builtin_function.go:120 +0x51 fp=0xc4237c5670 sp=0xc4237c5620
github.com/dop251/goja.(*Runtime).functionproto_call(0xc4f5bc0180, 0xe65040, 0xc4f5bf8ec0, 0xc4f5c06190, 0x4, 0x1e5, 0xb37edd, 0xc4237c5788)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/builtin_function.go:74 +0x157 fp=0xc4237c5708 sp=0xc4237c5670
github.com/dop251/goja.(*Runtime).(github.com/dop251/goja.functionproto_call)-fm(0xe65040, 0xc4f5bf8ec0, 0xc4f5c06190, 0x4, 0x1e5, 0xc4f5bbd620, 0x40b222)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/builtin_function.go:160 +0x51 fp=0xc4237c5758 sp=0xc4237c5708
github.com/dop251/goja.(*vm)._nativeCall(0xc4f4ced040, 0xc4f5bc4000, 0x4)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/vm.go:1827 +0x2e9 fp=0xc4237c5800 sp=0xc4237c5758
github.com/dop251/goja.call.exec(0xc400000004, 0xc4f4ced040)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/vm.go:1808 +0x6ff fp=0xc4237c58c8 sp=0xc4237c5800
github.com/dop251/goja.(*call).exec(0xc42110f840, 0xc4f4ced040)
	<autogenerated>:816 +0x55 fp=0xc4237c5908 sp=0xc4237c58c8
github.com/dop251/goja.(*vm).run(0xc4f4ced040)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/vm.go:288 +0x7f fp=0xc4237c5940 sp=0xc4237c5908
github.com/dop251/goja.(*funcObject).Call(0xc4f5ba7680, 0xe65040, 0xc4f5bf9d80, 0xc4f5c060b0, 0x3, 0x1f3, 0x10, 0xc4f5befae0)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/func.go:130 +0x324 fp=0xc4237c59e8 sp=0xc4237c5940
github.com/dop251/goja.(*funcObject).Call-fm(0xe65040, 0xc4f5bf9d80, 0xc4f5c060b0, 0x3, 0x1f3, 0x4085ce, 0xa45f40)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/builtin_function.go:120 +0x51 fp=0xc4237c5a38 sp=0xc4237c59e8
github.com/dop251/goja.(*Runtime).functionproto_call(0xc4f5bc0180, 0xe65040, 0xc4f5bc7400, 0xc4f5c060a0, 0x4, 0x1f4, 0xc4f5bbe5a0, 0xc4237c5b50)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/builtin_function.go:74 +0x157 fp=0xc4237c5ad0 sp=0xc4237c5a38
github.com/dop251/goja.(*Runtime).(github.com/dop251/goja.functionproto_call)-fm(0xe65040, 0xc4f5bc7400, 0xc4f5c060a0, 0x4, 0x1f4, 0xc4f5bbd620, 0x40b222)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/builtin_function.go:160 +0x51 fp=0xc4237c5b20 sp=0xc4237c5ad0
github.com/dop251/goja.(*vm)._nativeCall(0xc4f4ced040, 0xc4f5bc4000, 0x4)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/vm.go:1827 +0x2e9 fp=0xc4237c5bc8 sp=0xc4237c5b20
github.com/dop251/goja.call.exec(0xc400000004, 0xc4f4ced040)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/vm.go:1808 +0x6ff fp=0xc4237c5c90 sp=0xc4237c5bc8
github.com/dop251/goja.(*call).exec(0xc42110f840, 0xc4f4ced040)
	<autogenerated>:816 +0x55 fp=0xc4237c5cd0 sp=0xc4237c5c90
github.com/dop251/goja.(*vm).run(0xc4f4ced040)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/vm.go:288 +0x7f fp=0xc4237c5d08 sp=0xc4237c5cd0
github.com/dop251/goja.(*vm).(github.com/dop251/goja.run)-fm()
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/vm.go:375 +0x2a fp=0xc4237c5d20 sp=0xc4237c5d08
github.com/dop251/goja.(*vm).try(0xc4f4ced040, 0xc4237c5dc8, 0x0)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/vm.go:370 +0x119 fp=0xc4237c5db0 sp=0xc4237c5d20
github.com/dop251/goja.(*vm).runTry(0xc4f4ced040, 0xb60fc0)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/vm.go:375 +0x4f fp=0xc4237c5de8 sp=0xc4237c5db0
github.com/dop251/goja.(*Runtime).RunProgram(0xc4f5bc0180, 0xc422a2cd80, 0x0, 0x0, 0x0, 0x0)
	/Users/uppfinnarn/Projects/Go/src/github.com/dop251/goja/runtime.go:806 +0x212 fp=0xc4237c5e70 sp=0xc4237c5de8
github.com/loadimpact/k6/js.(*Bundle).instantiate(0xc423e0b680, 0xc4f5bc0180, 0xc4f5ba5d00, 0xc420157600, 0xc4f5baf700)
	/Users/uppfinnarn/Projects/Go/src/github.com/loadimpact/k6/js/bundle.go:219 +0xde fp=0xc4237c5f08 sp=0xc4237c5e70
github.com/loadimpact/k6/js.(*Bundle).Instantiate(0xc423e0b680, 0x0, 0x0, 0x0)
	/Users/uppfinnarn/Projects/Go/src/github.com/loadimpact/k6/js/bundle.go:195 +0x13c fp=0xc4237c5f80 sp=0xc4237c5f08
github.com/loadimpact/k6/js.(*Runner).newVU(0xc42124b780, 0xad55a0, 0x1, 0xc4f5ba5740)
	/Users/uppfinnarn/Projects/Go/src/github.com/loadimpact/k6/js/runner.go:102 +0x46 fp=0xc4237c60e0 sp=0xc4237c5f80
github.com/loadimpact/k6/js.(*Runner).NewVU(0xc42124b780, 0xc4f5ba5740, 0xc4f5bad6d0, 0x0, 0x0)
	/Users/uppfinnarn/Projects/Go/src/github.com/loadimpact/k6/js/runner.go:93 +0x2f fp=0xc4237c6110 sp=0xc4237c60e0
github.com/loadimpact/k6/core/local.(*Executor).SetVUsMax(0xc421cfb600, 0x1f40, 0x0, 0x0)
	/Users/uppfinnarn/Projects/Go/src/github.com/loadimpact/k6/core/local/local.go:433 +0x261 fp=0xc4237c61d0 sp=0xc4237c6110
github.com/loadimpact/k6/core.NewEngine(0xe659e0, 0xc421cfb600, 0x0, 0x1, 0x0, 0x1f40, 0x1, 0x0, 0x0, 0x0, ...)
	/Users/uppfinnarn/Projects/Go/src/github.com/loadimpact/k6/core/engine.go:83 +0x18c fp=0xc4237c63c0 sp=0xc4237c61d0
main.actionRun(0xc4221766e0, 0x0, 0xc4221766e0)
	/Users/uppfinnarn/Projects/Go/src/github.com/loadimpact/k6/run.go:466 +0x1891 fp=0xc4237c7820 sp=0xc4237c63c0
gopkg.in/urfave/cli%2ev1.HandleAction(0xa57880, 0xb61608, 0xc4221766e0, 0xc422bf8300, 0x0)
	/Users/uppfinnarn/Projects/Go/src/gopkg.in/urfave/cli.v1/app.go:490 +0xd4 fp=0xc4237c7848 sp=0xc4237c7820
gopkg.in/urfave/cli%2ev1.Command.Run(0xb376f9, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0, 0xb496d3, 0x1a, 0x0, ...)
	/Users/uppfinnarn/Projects/Go/src/gopkg.in/urfave/cli.v1/command.go:210 +0xb6e fp=0xc4237c7bf8 sp=0xc4237c7848
gopkg.in/urfave/cli%2ev1.(*App).Run(0xc420a69d40, 0xc42000e090, 0x9, 0x9, 0x0, 0x0)
	/Users/uppfinnarn/Projects/Go/src/gopkg.in/urfave/cli.v1/app.go:255 +0x76f fp=0xc4237c7de0 sp=0xc4237c7bf8
main.main()
	/Users/uppfinnarn/Projects/Go/src/github.com/loadimpact/k6/main.go:81 +0x6eb fp=0xc4237c7f88 sp=0xc4237c7de0
runtime.main()
	/usr/local/Cellar/go/1.8.3/libexec/src/runtime/proc.go:185 +0x20a fp=0xc4237c7fe0 sp=0xc4237c7f88
runtime.goexit()
	/usr/local/Cellar/go/1.8.3/libexec/src/runtime/asm_amd64.s:2197 +0x1 fp=0xc4237c7fe8 sp=0xc4237c7fe0

goroutine 5 [syscall]:
os/signal.signal_recv(0x0)
	/usr/local/Cellar/go/1.8.3/libexec/src/runtime/sigqueue.go:116 +0x104
os/signal.loop()
	/usr/local/Cellar/go/1.8.3/libexec/src/os/signal/signal_unix.go:22 +0x22
created by os/signal.init.1
	/usr/local/Cellar/go/1.8.3/libexec/src/os/signal/signal_unix.go:28 +0x4
```

**Comment by ragnarlonn** [comment]
Have you checked how much memory it uses on your Mac?  r4.large has 15 GB memory, and I have seen 1000 VU tests consuming several GB of memory. Also, it may be worth making sure you're running k6 v0.17.2 as it has some memory optimizations that makes it use less memory than previous versions.

**Comment by yzhang-myob** [comment]
@ragnarlonn Thank you for your reply.
First of all, I mess up with the concept of virtual users and request per seconds, which means I don't need 8000 vu anymore. But the problem is still there. 

In MAC, I can see the progress bar in k6, and memory usage increased overtime (until it hits the maximum file descriptor allowed). 
![screen shot 2017-10-03 at 07 51 20](https://user-images.githubusercontent.com/22581594/31099253-1d417e1c-a811-11e7-8134-23adb88b580a.png)

In Linux, I don't see the progress bar at all, and memory usage increased to 5G under 1 minute.
![screen shot 2017-10-03 at 08 00 03](https://user-images.githubusercontent.com/22581594/31099230-06803b96-a811-11e7-8f9a-63748db308bd.png)

Both machines have 16GB RAM. Not sure if this has anything to do with the ulimit on Amazon Linux.

**Comment by liclac** [comment]
k6 will preallocate the base environments for `-m`/`--max` VUs on startup, but each VU's local state (including HTTP responses) will occupy memory as well. Using 0.17.2 or later greatly reduces the rate at which memory usage will go up at runtime by recycling buffers for temporary objects.

If you wanna know how much memory just the VUs themselves take up, try running `k6 run --paused ...` and checking the memory usage of the k6 process.

**Comment by yzhang-myob** [comment]
@liclac First of all, I am sure that I am using k6 0.17.2 (although version is still shown as 0.17.1, I believe there is a bug ticket open for that). Secondly, I still observe the same behaviour.
On mac, paused with progress bar, and memory usage around 1.5G:
<img width="882" alt="screen shot 2017-10-04 at 07 12 09" src="https://user-images.githubusercontent.com/22581594/31147458-bc86d750-a8d5-11e7-824f-98410709e29b.png">

On Linux, paused without progress bar, and memory usage keep going up:
<img width="595" alt="screen shot 2017-10-04 at 07 26 23" src="https://user-images.githubusercontent.com/22581594/31147466-c17eec70-a8d5-11e7-8e79-ea2590f54529.png">

**Comment by liclac** [comment]
Just to be sure, could you build latest master on both machines? We're doing nothing different between different OSes apart from some minor details about how the CLI is rendered between *nix and windows.

**Comment by yzhang-myob** [comment]
@liclac Thanks for the suggestion. I found out the issue. I am actually running k6 0.71.1 on MAC since it is installed through home brew (Sorry for any confusion this may caused). If I use the master version, I can observe the same behaviour on MAC and Linux. Now the interesting question is what has changed between 0.71.1 and 0.71.2 to cause this? 

If this behaviour is desired, feel free to close this issue.

Thank you

**Comment by liclac** [comment]
Hm. https://github.com/loadimpact/k6/compare/v0.17.1...v0.17.2

The two big things I can think of are b002f2c107b7dbae486813647c44e2d215f5b336 and c9ab90b25e873af8ca2a4461d4396fcc9db346e6, the former should just reduce memory churn, the latter would increase memory usage by a bit... but it shouldn't be that bad?

**Comment by yzhang-myob** [comment]
Hi @liclac, sorry I am not a go developer to review the codes. But I did some experiment (hopefully I did it correctly...):
1. Check out k6 src, and create a branch from tag v0.17.1.
2. Cherry pick the cli fix commit (sorry, I don't know how to build without this commit)
3. Run `go get .` and `go install .`
4. run `./k6 run -u 8000 -p <my_script>`

I am observing the same issue as what I had with v0.17.2. I pretty sure the cli fix commit won't cause this. So perhaps some dependencies k6 using? I don't know what are the differences between `v0.17.1` was compiled for release and I compiled it now locally.

I definitely don't have this issue with v0.17.1 release binary.

Thanks for all your help and time!
